plagiarism/process.go

```diff
@@ -0,0 +1,136 @@
+package plagiarism
+
+import (
+    "io"
+    "io/ioutil"
+    "encoding/json"
+    "log"
+    "github.com/SpaceWhite/gomoss"
+    "fmt"
+    "os"
+)
+
+// GUIDsStruct use for decoding JSON
+type GUIDsStruct struct {
+    GUIDs []string
+}
+
+// Analysis struct contains output dir and reference dir
+type Analysis struct {
+    GIUDs []string
+    SubmissionDir string
+    Mossid uint32
+    Logger *log.Logger
+    ResultDir string
+}
+
+// DecodeRequest decode request from request body
+func DecodeRequest(body io.ReadCloser) ([]string, error) {
+    decoder := json.NewDecoder(body)
+    var guidsStruct GUIDsStruct
+    err := decoder.Decode(&guidsStruct)
+
+    if err != nil {
+        return []string{}, err
+    }
+    return guidsStruct.GUIDs, nil
+}
+
+// GUIDPath use to get submission from OmegaUp
+// Example GUID : abcdefghi1234
+// GUID Path    : submissionDir/ab/cdefghi1234
+func (analysis *Analysis) GUIDPath(guid string) string {
+    return fmt.Sprintf("%v/%v/%v", analysis.SubmissionDir, guid[:2], guid[2:])
+}
+
+// GUIDResultDir use to specify in what directory the analysis results are stored
+// Example GUID : abcdefghi1234
+// GUID Path    : resultDir/ab
+func (analysis *Analysis) GUIDResultDir(guid string) string {
+    return fmt.Sprintf("%v/%v", analysis.ResultDir, guid[:2])
+}
+
+// GUIDResultPath use to specify in what path the analysis results are stored
+// Example GUID : abcdefghi1234
+// GUID Path    : resultDir/abcdefghi1234
+func (analysis *Analysis) GUIDResultPath(guid string) string {
+    return fmt.Sprintf("%v/%v/%v", analysis.ResultDir, guid[:2], guid[2:])
+}
+
+// RunDetector requests MOSS server to run plagiarism detector
+func (analysis *Analysis) RunDetector() Response {
```

Comment:
This is going to take a loooong time. So it's probably going to be better if you run this in a goroutine:

```go
func (analyzer *Analyzer) StartPlagiarismDetection(...) (id string) {
  analyzer.queue <- ...
  return analyzer.createId()
}

func (analyzer *Analyzer) Shutdown(ctx context.Context) error {
  analyzer.queue <- nil
  for {
    select {
    case <-analyzer.done:
      return nil
    case <-ctx.Done():
      return ctx.Err()
    }
  }
}

func (analyzer *Analyzer) RunLoop() {
  for {
    select {
    case request := <-analyzer.queue:
      if (request == nil) {
        analyzer.ctx.Log.Info("nil request received, stopping run loop")
        break
      }
      analyzer.ProcessOne(request)
    }
  }
  close(analyzer.done)
}

...

// in main()
analyzer := plagiarism.NewAnalyzer(...)
go analyzer.RunLoop()
```